ocean_lib/ocean/ocean_exchange.py

```diff
@@ -82,10 +82,23 @@ def buy_at_fixed_rate(
                 f"tokens which exceeds the max_OCEAN_amount {max_OCEAN_amount}."
             )
         ocean_token = DataToken(self._web3, self.ocean_address)
-        ocean_token.get_tx_receipt(
-            self._web3,
-            ocean_token.approve(self._exchange_address, ocean_amount_base, wallet),
-        )
+        if (
+            ocean_token.allowance(wallet.address, self._exchange_address)
+            < ocean_amount_base
+        ):
+            tx_id = ocean_token.approve(
+                self._exchange_address, ocean_amount_base, wallet
+            )
+            tx_receipt = ocean_token.get_tx_receipt(
+                self._web3,
+                tx_id,
+            )
+            if not tx_receipt or tx_receipt.status != 1:
+                raise VerifyTxFailed(
+                    f"Approve OCEAN tokens failed, exchange address was {self._exchange_address} and tx id was {tx_id}!"
+                )
+        if ocean_token.balanceOf(wallet.address) < amount_base:
```

Comment:
Two things:

1. This logic doesn't make sense.  I suspect that `amount_base` is meant to be `ocean_amount_base`
2. If I'm correct about (1), then I think this logic should be performed before the approval but after `max_OCEAN_amount` check.

Like so:

```python
        if ocean_amount_base > max_OCEAN_amount_base:
            raise ValidationError(
                f"Buying {amount} datatokens requires {from_base_18(ocean_amount_base)} OCEAN "
                f"tokens which exceeds the max_OCEAN_amount {max_OCEAN_amount}."
            )
        ocean_token = DataToken(self._web3, self.ocean_address)
        if ocean_token.balanceOf(wallet.address) < ocean_amount_base:
            raise InsufficientBalance(f"Insufficient funds for buying {amount_base} DataTokens!")
```

P.S. I added `{amount_base}` to the InsufficientBalance error message.